### DISCUSSION TITLE
support enums in role names

### DISCUSSION THREAD

**Discussion by ceejayoz** [spatie/laravel-permission#1991] [Ideas]
We have a PHP 8.1 BackedEnum listing our application's roles.

This permits us to do the following:

    Spatie\Permission\Models\Role::create(['name' => App\Enums\Role::ADMINISTRATOR]);

However, this does not work with `assignRole()`:

    $user->assignRole(App\Enums\Role::ADMINISTRATOR)

and we are instead required to do:

    $user->assignRole(App\Enums\Role::ADMINISTRATOR->value)

This appears to be due to [HasRoles::getStoredRole accepting a string, numeric, or Role as argument](https://github.com/spatie/laravel-permission/blob/6a3ed627cee28a552b5176c172ae0abc5eb30925/src/Traits/HasRoles.php#L325).

I'd suggest this function have an additional check along the lines of:

        if (is_object($role) && $role instanceof \BackedEnum) {
            return $roleClass->findByName($role->value, $this->getDefaultGuardName());
        }

so enums can be used as role names.

**Comment by erikn69** [comment]
Did you try overriding `getStoredRole` on your User model implementation?
```php
    use \Spatie\Permission\Traits\HasRoles {
        getStoredRole as getStoredRoleTrait;
    }
    protected function getStoredRole($role): Role
    {
        if (is_object($role) && $role instanceof \BackedEnum) {
              $roleClass = $this->getRoleClass();
              return $roleClass->findByName($role->value, $this->getDefaultGuardName());
        }

        return $this->getStoredRoleTrait($role);
    }
```

**Comment by edalzell** [comment]
To make it work w/ `hasRole` you also need to completely override `hasRole` and add:

```php
if ($roles instanceof BackedEnum) {
            return $guard
                ? $this->roles->where('guard_name', $guard)->contains('name', $roles->value)
                : $this->roles->contains('name', $roles->value);
        }
```

**Reply by edalzell** [reply]
This is worth a PR, I think

**Reply by drbyte** [reply]
See #2391

**Comment by boumanb** [comment]
Just wanted to let you know I had the same idea and solved it like this:

```php
<?php

namespace App\Policies;

use  App\Models\Role as RoleModel;

enum Role: string
{
    case LEAD_ADMIN = 'lead_admin';
    case GLOBAL_USER = 'global_user';
    case ADMINISTRATOR = 'administrator';
    case DEVELOPER = 'developer';
    case PUBLISHER = 'publisher';
    case AUTHOR = 'author';

    public function id(): int
    {
        return RoleModel::whereName($this->value)->select('id')->first()->id;
    }
}
```

```php
<?php

namespace App\Concerns;

use App\Policies\Role;

trait HasRoles
{
    use \Spatie\Permission\Traits\HasRoles {
        assignRole as protected spatieAssignRole;
    }

    public function assignRole(Role $role)
    {
        $this->spatieAssignRole($role->value);

        return $this;
    }
}
```
And in the `App\Models\User`:
```php
<?php

namespace App\Models;

use App\Concerns\HasRoles;

class User extends Authenticatable
{
    use HasRoles;
    .....
}
```

**Reply by drbyte** [reply]
Feedback welcome: #2391

**Reply by silnex** [reply]
I use a way with fewer breakpoints.

```php
<?php

namespace App\Models\Trait;

trait HasRoles
{
    use \Spatie\Permission\Traits\HasRoles {
        assignRole as protected spatieAssignRole;
    }

    public function assignRole(...$roles)
    {
        $roles = collect($roles)->map(function ($role) {
            if ($role instanceof \BackedEnum) {
                return $role->value;
            }

            return $role;
        });

        $this->spatieAssignRole(...$roles);

        return $this;
    }
}
```

**Reply by drbyte** [reply]
@silnex thanks for your note. 
I realized that I missed a few things in the initial implementation in #2391; it's been updated in https://github.com/spatie/laravel-permission/commit/299fe4beb7e6f77079c9642e92bbd887d8736b6f

**Comment by JacobHonore** [comment]
Is it a mistake that the Enums-feature is currently shown in the v5 documentation? It seems to be a v6 feature
https://spatie.be/docs/laravel-permission/v5/basic-usage/enums

**Reply by drbyte** [reply]
> Is it a mistake that the Enums-feature is currently shown in the v5 documentation? It seems to be a v6 feature
> https://spatie.be/docs/laravel-permission/v5/basic-usage/enums

Fixed.

**Comment by localusercamp** [comment]
Will there be support for this feature?

This is very ugly =D
![image](https://github.com/spatie/laravel-permission/assets/45385480/ac9a9f47-66ae-40e9-a0b1-887656cf8efb)

**Reply by erikn69** [reply]
Already on v6: 

- https://github.com/spatie/laravel-permission/pull/2391
